### PR TITLE
add batch impl. for inplace `index_add` operation

### PR DESCRIPTION
**Pull request by guilhermeleobas**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112276

cc @zou3519 @Chillee @samdow @kshitij12345 @janeyx99

### CODE REVIEW

**Comment by guilhermeleobas** [comment]
ref: https://github.com/pytorch/pytorch/issues/105539

File: test/functorch/test_ops.py [hunk 1]
```diff
@@ -956,6 +955,7 @@ def vjp_of_vjp(*args_and_cotangents):
              {torch.float32: tol(atol=5e-04, rtol=1e-04)}, device_type="cuda"),
     ))
     @skipOps('TestOperators', 'test_vmapvjp', vmapvjp_fail.union({
+        xfail('as_strided'),
```
Comment (kshitij12345) [inline test/functorch/test_ops.py:958]: We are just moving above xfail here? Is that correct?

File: aten/src/ATen/functorch/BatchRulesScatterOps.cpp [hunk 2]
```diff
@@ -977,11 +991,40 @@ std::tuple<Tensor,optional<int64_t>> index_add_batch_rule(
       other.select(*other_bdim, i) : other;
     const auto& index_slice = index_bdim.has_value() ?
       index.select(*index_bdim, i) : index;
-    results.push_back(at::index_add(self_slice, dim, index_slice, other_slice, alpha));
+
+    if (inplace) {
```
Comment (kshitij12345) [inline aten/src/ATen/functorch/BatchRulesScatterOps.cpp:995]: Is there a simple way to not initialize `results` for inplace case (as it is not used then).

Or maybe we should guard call to reserve with `if (!inplace) { results.reserver(batch_size)}`

Wdyt?

**Review by kshitij12345 — COMMENTED** [review]
Looks good but have couple of questions. Thank you!

File: test/functorch/test_ops.py (hunk 1 above)
Comment (guilhermeleobas) [reply test/functorch/test_ops.py:958]: Yes, that's correct. Before the xfail was in `vmapvjp_fail`, which is used in other places of this file. With the addition of `index_add`, there is a batch rule for `as_strided`, but it fails on some specific cases.

File: aten/src/ATen/functorch/BatchRulesScatterOps.cpp (hunk 2 above)
Comment (guilhermeleobas) [reply aten/src/ATen/functorch/BatchRulesScatterOps.cpp:997]: Done! I've guarded the initialization as you suggested.

**Review by zou3519 — APPROVED** [review]

**Review by kshitij12345 — APPROVED** [review]
LGTM, thanks @guilhermeleobas

**Comment by PaliC** [comment]
@pytorchbot revert -m "breaking linux binary builds" -c "nosignal"

You can find the breakage here: https://hud.pytorch.org/pytorch/pytorch/commit/e3c8c63deaf594699d827e84869a3ecd7e2ab494

**Comment by guilhermeleobas** [comment]
Hi @PaliC, failure doesn't seem to be related to the changes I did in this PR:

```
...
2023-10-31T04:53:40.4122964Z ++ echo 'Copied /usr/lib64/libgomp.so.1 to torch/lib/libgomp-a34b3233.so.1'
2023-10-31T04:53:40.4123779Z Copied /usr/lib64/libgomp.so.1 to torch/lib/libgomp-a34b3233.so.1
2023-10-31T04:53:40.4124389Z ++ for filepath in '"${DEPS_LIST[@]}"'
2023-10-31T04:53:40.4124860Z +++ basename /usr/local/cuda/lib64/libcusparseLt.so.0
2023-10-31T04:53:40.4133181Z ++ filename=libcusparseLt.so.0
2023-10-31T04:53:40.4133594Z ++ destpath=torch/lib/libcusparseLt.so.0
2023-10-31T04:53:40.4134414Z ++ [[ /usr/local/cuda/lib64/libcusparseLt.so.0 != \t\o\r\c\h\/\l\i\b\/\l\i\b\c\u\s\p\a\r\s\e\L\t\.\s\o\.\0 ]]
2023-10-31T04:53:40.4135402Z ++ cp /usr/local/cuda/lib64/libcusparseLt.so.0 torch/lib/libcusparseLt.so.0
2023-10-31T04:53:40.4143847Z cp: cannot stat ‘/usr/local/cuda/lib64/libcusparseLt.so.0’: No such file or directory
```

Can I just rebase and merge it again?

**Review by malfet — APPROVED** [review]

**Comment by malfet** [comment]
> @pytorchbot revert -m "breaking linux binary builds" -c "nosignal"
> 
> You can find the breakage here: https://hud.pytorch.org/pytorch/pytorch/commit/e3c8c63deaf594699d827e84869a3ecd7e2ab494

I think failure is unrelated, as revert clearly did not help, likely cause of regression is https://github.com/pytorch/builder/commit/e9ce243cf5174035a68ec23d70156be7399ebd94 that modified docker images

**Pull request closed without merging**